libs/utils/src/lock_file.rs

```diff
@@ -0,0 +1,77 @@
+//! A module to allow processes to lock on specific FS file exclusive.
+//! Could be used to forbid multiple processes to start in the same directory.
+//!
+//! Note that locked files are done via [`fcntl::flock`] which means that
+//! * files are locked only for taking another lock on them: all other operations are still possible on files.
+//! For example, other process could do `rm file` and that would be allowed.
+//! * if no file exists (after `rm` in another process), the process that "locked" the file still runs and does not error
+//! new procesess are able to start in the same directory, as they will be creating another file to lock.
+
+use std::{
+    fs,
+    os::unix::prelude::{AsRawFd, OpenOptionsExt},
+    path::Path,
+};
+
+use anyhow::Context;
+use nix::{fcntl, libc};
+
+use crate::crashsafe;
+
+pub enum LockCreationResult {
+    Created {
+        new_lock_contents: String,
+        file: fs::File,
+    },
+    AlreadyLocked {
+        existing_lock_contents: String,
+    },
+    CreationFailed(anyhow::Error),
+}
+
+/// Creates a lock file in the path given and writes the given contents into the file.
+pub fn create_lock_file<F>(lock_file: &Path, contents_producer: F) -> LockCreationResult
+where
+    F: FnOnce() -> String,
+{
+    let file = match fs::OpenOptions::new()
+        .create(true) // O_CREAT
+        .write(true)
+        .custom_flags(libc::O_CLOEXEC)
+        .open(lock_file)
+        .context("Failed to open lock file")
+    {
+        Ok(file) => file,
+        Err(e) => return LockCreationResult::CreationFailed(e),
+    };
+
+    match fcntl::flock(file.as_raw_fd(), fcntl::FlockArg::LockExclusiveNonblock) {
+        Ok(()) => {
+            let new_lock_contents = contents_producer();
+            match fs::write(lock_file, &new_lock_contents)
+                .with_context(|| {
+                    format!("Failed to write '{new_lock_contents}' contents into lockfile")
+                })
+                .and_then(|()| {
+                    crashsafe::fsync_file_and_parent(lock_file).context("Failed to fsync lockfile")
```

Comment:
Fsyncing is unnecessary here. If you lose power or the kernel crashes or similar, the process will go away along with the system, we don't care about the pidfile anymore